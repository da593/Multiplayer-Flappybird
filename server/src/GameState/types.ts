export interface Coordinates {
    x: number;
    y: number;
}

export interface BoxCoordinates {
    topLeft: Coordinates;
    topRight: Coordinates;
    botLeft: Coordinates;
    botRight: Coordinates;
}

export interface GameState_I {
    gapCoords: BoxCoordinates;
    birdCoords: BoxCoordinates;
    score: number;
    hasCollided: boolean;
}

export interface Dimensions_I {
    GAME_WIDTH: number,
    GAME_HEIGHT: number,
    PIPE_WIDTH: number,
    GAP_HEIGHT: number,
    PIPE_VELOCITY: number,
    BIRD_WIDTH: number,
    BIRD_X_LOCATION: number,
    Y_FLY_UP: number,
}