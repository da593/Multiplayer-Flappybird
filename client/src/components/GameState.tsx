import React, {useState,useEffect} from "react";
import { BoardGame } from "./BoardGame";
import * as consts from "components/GameConstants";

export interface Coordinates {
    x: number;
    y: number;
}

export interface BoxCoordinates {
    topLeft: Coordinates;
    topRight: Coordinates;
    botLeft: Coordinates;
    botRight: Coordinates;
}

export function GameState() {
    const [xPipeLoc,setXPipeLoc] = useState(consts.GAME_WIDTH + consts.PIPE_WIDTH);
    const [yGapLoc,setYGapLoc] = useState(randomizeYGapLocation());
    const [gapCoords,setGapCoords] = useState<BoxCoordinates>({topLeft: {x: 0, y:0}, topRight: {x: 0, y:0}, botLeft: {x: 0, y:0}, botRight: {x: 0, y:0}});

    const [yBirdLoc,setYBirdLoc] = useState(consts.GAP_HEIGHT / 3);
    const [birdCoords,setBirdCoords] = useState<BoxCoordinates>({topLeft: {x: 0, y:0}, topRight: {x: 0, y:0}, botLeft: {x: 0, y:0}, botRight: {x: 0, y:0}});

    const [hasCollided,setHasCollided] = useState(false);
    const [hasKeyClicked,setHasKeyClicked] = useState(false);

    function randomizeYGapLocation():number {
        return Math.floor(Math.random() * (consts.GAME_WIDTH - consts.GAP_HEIGHT));
    }

    const detectCollision = () => {
        if (birdCoords.botLeft.y >= consts.GAME_HEIGHT) {
            setHasCollided(true);
        }
    }   

    const animatePipe = (frameCount:number) => {
        if (frameCount == 0) {
            setYGapLoc(randomizeYGapLocation());
            setXPipeLoc(consts.GAME_WIDTH + consts.PIPE_WIDTH);
        }
        else {
            setXPipeLoc(xPipeLoc - frameCount);
        }  
    }

    const animateBird = (frameCount:number,isEvent=false) => {
        if (hasKeyClicked == true) {
            setYBirdLoc(yBirdLoc - 40);
            setHasKeyClicked(false);
        }
        else if (frameCount == 0) {
            setYBirdLoc(consts.GAP_HEIGHT / 3)
        }
        else {
            setYBirdLoc(yBirdLoc + frameCount);
        }
    }

    const handleClick = (e:KeyboardEvent) => {
        if (e.key === "w") {
            setHasKeyClicked(true);
        }
    }    
    useEffect(() => {
        let raf:number;
        let frameCount = 0;
        const render = () => {
            if (!hasCollided) {
                frameCount == consts.GAME_WIDTH + consts.PIPE_WIDTH ? frameCount = 0 : frameCount++;
                animatePipe(frameCount)
                raf = window.requestAnimationFrame(render);
            }
        }
        render();
        return () => {
            window.cancelAnimationFrame(raf);
        }
    },[hasCollided])

    useEffect(() => {
        window.addEventListener("keydown",handleClick)
        let raf:number;
        let frameCount = 0;
        const render = () => {
            if (!hasCollided) {
                frameCount++
                animateBird(frameCount);
                raf = window.requestAnimationFrame(render);
            }
        }
        render();
        return () => {
            window.cancelAnimationFrame(raf);
        }
    },[hasCollided,hasKeyClicked])
    
    useEffect(() => {
        const gapTopLeft:Coordinates = {x:xPipeLoc - consts.PIPE_WIDTH, y:yGapLoc - consts.GAP_HEIGHT};
        const gapTopRight:Coordinates = {x:xPipeLoc, y:yGapLoc - consts.GAP_HEIGHT};
        const gapBotLeft:Coordinates = {x:xPipeLoc - consts.PIPE_WIDTH, y:yGapLoc};
        const gapBotRight:Coordinates = {x:xPipeLoc, y:yGapLoc};
        setGapCoords({topLeft:gapTopLeft,topRight: gapTopRight, botLeft: gapBotLeft, botRight: gapBotRight})

        const birdTopLeft:Coordinates = {x: consts.BIRD_X_LOCATION - consts.BIRD_WIDTH, y: yBirdLoc - consts.BIRD_WIDTH};
        const birdTopRight:Coordinates = {x: consts.BIRD_X_LOCATION, y: yBirdLoc - consts.BIRD_WIDTH};
        const birdBotLeft:Coordinates = {x: consts.BIRD_X_LOCATION - consts.BIRD_WIDTH, y: yBirdLoc};
        const birdBotRight:Coordinates = {x: consts.BIRD_X_LOCATION, y: yBirdLoc};

        setBirdCoords({topLeft:birdTopLeft,topRight:birdTopRight, botLeft:birdBotLeft, botRight:birdBotRight})
    },[xPipeLoc,yBirdLoc])

    useEffect(() => {
        detectCollision()
    },[gapCoords,birdCoords])


    return (
        <BoardGame width={consts.GAME_WIDTH} height={consts.GAME_HEIGHT} xPipeLoc={xPipeLoc} yGapLoc={yGapLoc} yBirdLoc={yBirdLoc}/>
    )
}