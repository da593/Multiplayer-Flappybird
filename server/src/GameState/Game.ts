import { BoxCoordinates, GameStateResponse, PlayerState_I, ReadyCheck, WinState, detectCollision, game_tick } from "@flappyblock/shared";
import { Player } from "#@/entities/Player.js";
import { Pipe } from "#@/entities/Pipe.js";

export class Game {
	players: Array<Player>
	pipe: Pipe;
	deadPlayers: Set<string>;
	lobbyCheck: Set<string>;
	hasStarted: boolean;
	shouldEnd: boolean;
	winner: string;

	constructor(pipe: Pipe) {
		this.players = new Array();
		this.pipe = pipe;
		this.deadPlayers = new Set();
		this.lobbyCheck = new Set();
		this.hasStarted = false;
		this.shouldEnd = false;
		this.winner = "none";
	}

 	getNumPlayers(): number {
		return this.players.length;
	}

	getWinner(): string {
		return this.winner;
	}

	getShouldEnd(): boolean {
		return this.shouldEnd;
	}

	getHasStarted(): boolean {
		return this.hasStarted;
	}

	getState(): GameStateResponse {
		const playerState: Record<string, PlayerState_I> = {};
		this.players.forEach((player: Player) => {
			playerState[player.getEntityId()] = player.getPlayerState();
		})

		return {
			players: playerState,
			pipe: this.pipe.getState(),
		};
	}

	addPlayer(newPlayer: Player): void {
		this.players.push(newPlayer);
	}

	removePlayer(playerId: string): void {
        const index = this.players.findIndex((elem:Player) => elem.getEntityId() === playerId);
        this.players.splice(index,1);
		this.deadPlayers.delete(playerId);
		this.lobbyCheck = new Set();
	}

	resetState(): void {
		this.players.forEach((player: Player) => {
			player.resetState()
		});
		this.pipe.resetState();
	}

	startGame(playerId: string): ReadyCheck {
		this.lobbyCheck.add(playerId);
		const readyCheck = {
			numReady: this.lobbyCheck.size,
		}
		if (readyCheck.numReady === this.players.length) {
			this.lobbyCheck = new Set();
			this.hasStarted = true;
			this.resetState();
			this.updateGame();
		}
		return readyCheck;
	}

	restartGame(playerId: string): ReadyCheck {
		this.lobbyCheck.add(playerId);
		const readyCheck = {
			numReady: this.lobbyCheck.size,
		}
		if (readyCheck.numReady ===  this.players.length) {
			this.deadPlayers = new Set();
			this.lobbyCheck = new Set();
			this.hasStarted = false;
			this.shouldEnd = false;
			this.winner = "none";
			this.resetState();
		}
		return readyCheck;
	}

	updateGame(): void {
		const gameLoopId = setInterval(() => {
			if (this.getNumPlayers() <= 0 || this.deadPlayers.size === this.players.length) {
				this.determineWinner();
				clearInterval(gameLoopId);
				this.shouldEnd = true;
			}
			this.pipe.update();
			this.players.forEach((player: Player) => {
				if (player.getHasCollided()) {
					this.deadPlayers.add(player.getEntityId());
				}
				else {
					player.update();
					const currBirdCoords = player.getBirdCoords();
					const currPipeCoords = this.pipe.getPipeCoords();
					player.setHasCollided(detectCollision(currBirdCoords, currPipeCoords));
					if (this.addScore(currBirdCoords, currPipeCoords)) {
						player.addScore();
					}
					
				}
			})
			
		}, game_tick);
	}

	playerInput(playerId: string): void {
		const player = this.players.find((player: Player) => player.getEntityId() === playerId);
		player?.userInput();
	}

	determineWinner(): void {
		if (this.players.length > 1) {
			let maxScore = this.players[0].score;
			this.winner = this.players[0].getEntityId();
			for (let i = 1; i < this.players.length; i++) {
				if (this.players[i].score > maxScore) {
					maxScore =  this.players[0].score;
					this.winner = this.players[i].getEntityId();
				}
				else if (this.players[i].score === maxScore) {
					this.winner = WinState.DRAW
				}
			}
		}
		else {
			this.winner = WinState.NO_WINNER
		}
	}

	setShouldEnd(bool: boolean) {
		this.shouldEnd = bool;
	}

	addScore(birdCoords: BoxCoordinates, pipeCoords: BoxCoordinates): boolean {
        if (birdCoords.topLeft.x === pipeCoords.topRight.x + 1) {
            return true;
        }
		return false;
    }

}