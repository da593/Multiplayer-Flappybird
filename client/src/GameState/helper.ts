import { GAME_WIDTH, PIPE_WIDTH, GAP_HEIGHT, GAME_HEIGHT, BIRD_X_LOCATION, BIRD_WIDTH } from "./constants";
import { Coordinates,BoxCoordinates } from "./types";

export const randomizeYGapLocation = ():number => {
    return Math.floor(Math.random() * (GAME_HEIGHT - GAP_HEIGHT));
}

export const detectCollision = (birdCoords:BoxCoordinates, gapCoords:BoxCoordinates):boolean => {
    return birdCoords.botLeft.y >= GAME_HEIGHT ||
    birdCoords.topRight.x >= gapCoords.topLeft.x && birdCoords.topLeft.x <= gapCoords.topRight.x && (
    birdCoords.topLeft.y <= gapCoords.topLeft.y ||
    birdCoords.botLeft.y >= gapCoords.botLeft.y)
    
}   

export const calculateGapCoords = (xPipeLoc: number, yGapLoc: number):BoxCoordinates => {
    const gapTopLeft:Coordinates = {x:xPipeLoc, y:yGapLoc};
    const gapTopRight:Coordinates = {x:xPipeLoc + PIPE_WIDTH, y:yGapLoc};
    const gapBotLeft:Coordinates = {x:xPipeLoc, y:yGapLoc + GAP_HEIGHT};
    const gapBotRight:Coordinates = {x:xPipeLoc + PIPE_WIDTH, y:yGapLoc + GAP_HEIGHT};
    return {topLeft:gapTopLeft,topRight: gapTopRight, botLeft: gapBotLeft, botRight: gapBotRight}
}

export const calculateBirdCoords = (yBirdLoc: number):BoxCoordinates => {
    const birdTopLeft:Coordinates = {x: BIRD_X_LOCATION, y: yBirdLoc};
    const birdTopRight:Coordinates = {x: BIRD_X_LOCATION + BIRD_WIDTH, y: yBirdLoc};
    const birdBotLeft:Coordinates = {x: BIRD_X_LOCATION, y: yBirdLoc + BIRD_WIDTH};
    const birdBotRight:Coordinates = {x: BIRD_X_LOCATION + BIRD_WIDTH, y: yBirdLoc + BIRD_WIDTH};
    return {topLeft:birdTopLeft,topRight: birdTopRight, botLeft: birdBotLeft, botRight: birdBotRight}
}

export const calculateNewGapCoords = (prevGapCoords: BoxCoordinates, frameCount:number):BoxCoordinates => {
    let xPipeLoc:number = prevGapCoords.topLeft.x;
    let yGapLoc: number = prevGapCoords.topLeft.y;

    if (xPipeLoc <= -PIPE_WIDTH) {
        yGapLoc = randomizeYGapLocation();
        xPipeLoc = GAME_WIDTH;
    }
    else {
        xPipeLoc = prevGapCoords.topLeft.x - frameCount;
    }
    
    return calculateGapCoords(xPipeLoc,yGapLoc);
}

export const calculateNewBirdCoords = (prevGapCoords: BoxCoordinates, frameCount:number, hasKeyClicked:boolean) => {
    let yBirdLoc = prevGapCoords.topLeft.y;
    if (hasKeyClicked === true && yBirdLoc > 0) {
        yBirdLoc = yBirdLoc - 40;
    }
    else {
        yBirdLoc = yBirdLoc + frameCount;
    }
    return calculateBirdCoords(yBirdLoc);
}



