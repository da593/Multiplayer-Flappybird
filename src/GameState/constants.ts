import { calculateBirdCoords, calculateGapCoords, randomizeYGapLocation } from "./helper";
import { Dimensions_I, InitalState_I } from "./types";

export const GAME_DIMENSIONS:Dimensions_I = {
    GAME_WIDTH: 480,
    GAME_HEIGHT: 640,
    PIPE_WIDTH: 100,
    GAP_HEIGHT: 125,
    PIPE_VELOCITY: -0.05,
    BIRD_WIDTH: 50,
    BIRD_X_LOCATION: 50,
    Y_FLY_UP: 25,
};

export  const INITAL_STATE:InitalState_I =  {
    gapCoords: calculateGapCoords(GAME_DIMENSIONS.GAME_WIDTH, randomizeYGapLocation()),
    birdCoords: calculateBirdCoords(GAME_DIMENSIONS.GAP_HEIGHT / 3),
    hasCollided: false,
    hasKeyClicked: false,
    score: 0,
}