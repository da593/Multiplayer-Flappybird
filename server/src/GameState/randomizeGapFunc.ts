import { GAME_DIMENSIONS } from "./constants";


export const randomizeYGapLocation = ():number => {
    return Math.floor(Math.random() * (GAME_DIMENSIONS.GAME_HEIGHT - GAME_DIMENSIONS.GAP_HEIGHT));
}