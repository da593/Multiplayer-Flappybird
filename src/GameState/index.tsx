import React, {useState,useEffect} from "react";
import { BoardGame } from "../GameBoard";
import {calculateNewBirdCoords, calculateNewGapCoords, detectCollision} from "./helper"
import { BoxCoordinates } from "./types";
import { INITAL_STATE } from "./constants";


export function GameState() {
    

    const [hasCollided,setHasCollided] = useState(INITAL_STATE.hasCollided);
    const [gapCoords,setGapCoords] = useState<BoxCoordinates>(INITAL_STATE.gapCoords);
    const [birdCoords,setBirdCoords] = useState<BoxCoordinates>(INITAL_STATE.birdCoords);
    const [hasKeyClicked,setHasKeyClicked] = useState(INITAL_STATE.hasKeyClicked);
    const [score,setScore] = useState(INITAL_STATE.score);
    

    const handleKeyBoardEvent = (e:KeyboardEvent) => {
        if (e.key === "w") {
            setHasKeyClicked(true);
        }
    }  

    const handleReset = () => {
        setGapCoords(INITAL_STATE.gapCoords)
        setBirdCoords(INITAL_STATE.birdCoords);
        setHasCollided(INITAL_STATE.hasCollided);
        setHasKeyClicked(INITAL_STATE.hasKeyClicked);
        setScore(INITAL_STATE.score);
    }

    useEffect(() => {
        window.addEventListener("keydown",handleKeyBoardEvent)
        let raf:number;
        const frameCount = 1;
        const render = () => {
            if (!hasCollided) {
                setBirdCoords(prevBirdCoords => calculateNewBirdCoords(prevBirdCoords,frameCount,hasKeyClicked))
                setGapCoords(prevGapCoords => calculateNewGapCoords(prevGapCoords,frameCount));
                setHasKeyClicked(false);
                raf = window.requestAnimationFrame(render);
            }
        }
        render();
        return () => {
            window.cancelAnimationFrame(raf);
        }
    },[hasCollided,hasKeyClicked])


    useEffect(() => {
        if (detectCollision(birdCoords,gapCoords)) {
            setHasCollided(true);
        }
        else if (birdCoords.topLeft.x === gapCoords.topRight.x + 1 && !hasCollided) {
            setScore(score + 1)
        }
    },[gapCoords,birdCoords])


    return (
        <BoardGame gapCoords={gapCoords} birdCoords={birdCoords} score={score} hasCollided={hasCollided} handleReset={handleReset}/>
    )
}

