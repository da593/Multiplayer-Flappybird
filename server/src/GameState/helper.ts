import { Coordinates,BoxCoordinates } from "./types";
import { GAME_DIMENSIONS } from "./constants";
import { randomizeYGapLocation } from "./randomizeGapFunc";


export const detectCollision = (birdCoords:BoxCoordinates, gapCoords:BoxCoordinates):boolean => {
    return birdCoords.botLeft.y >= GAME_DIMENSIONS.GAME_HEIGHT ||
    birdCoords.topRight.x >= gapCoords.topLeft.x && birdCoords.topLeft.x <= gapCoords.topRight.x && (
    birdCoords.topLeft.y <= gapCoords.topLeft.y ||
    birdCoords.botLeft.y >= gapCoords.botLeft.y)
    
}   

export const calculateGapCoords = (xPipeLoc: number, yGapLoc: number):BoxCoordinates => {
    const gapTopLeft:Coordinates = {x:xPipeLoc, y:yGapLoc};
    const gapTopRight:Coordinates = {x:xPipeLoc + GAME_DIMENSIONS.PIPE_WIDTH, y:yGapLoc};
    const gapBotLeft:Coordinates = {x:xPipeLoc, y:yGapLoc + GAME_DIMENSIONS.GAP_HEIGHT};
    const gapBotRight:Coordinates = {x:xPipeLoc + GAME_DIMENSIONS.PIPE_WIDTH, y:yGapLoc + GAME_DIMENSIONS.GAP_HEIGHT};
    return {topLeft:gapTopLeft,topRight: gapTopRight, botLeft: gapBotLeft, botRight: gapBotRight}
}

export const calculateBirdCoords = (yBirdLoc: number):BoxCoordinates => {
    const birdTopLeft:Coordinates = {x: GAME_DIMENSIONS.BIRD_X_LOCATION, y: yBirdLoc};
    const birdTopRight:Coordinates = {x: GAME_DIMENSIONS.BIRD_X_LOCATION + GAME_DIMENSIONS.BIRD_WIDTH, y: yBirdLoc};
    const birdBotLeft:Coordinates = {x: GAME_DIMENSIONS.BIRD_X_LOCATION, y: yBirdLoc + GAME_DIMENSIONS.BIRD_WIDTH};
    const birdBotRight:Coordinates = {x: GAME_DIMENSIONS.BIRD_X_LOCATION + GAME_DIMENSIONS.BIRD_WIDTH, y: yBirdLoc + GAME_DIMENSIONS.BIRD_WIDTH};
    return {topLeft:birdTopLeft,topRight: birdTopRight, botLeft: birdBotLeft, botRight: birdBotRight}
}

export const calculateNewGapCoords = (prevGapCoords: BoxCoordinates):BoxCoordinates => {
    let xPipeLoc:number = prevGapCoords.topLeft.x;
    let yGapLoc: number = prevGapCoords.topLeft.y;
    if (xPipeLoc <= -GAME_DIMENSIONS.PIPE_WIDTH) {
        yGapLoc = randomizeYGapLocation();
        xPipeLoc = GAME_DIMENSIONS.GAME_WIDTH;
    }
    else {
        xPipeLoc = prevGapCoords.topLeft.x - 1;
    }
    
    return calculateGapCoords(xPipeLoc,yGapLoc);
}

export const calculateNewBirdCoords = (prevGapCoords: BoxCoordinates) => {
    let yBirdLoc = prevGapCoords.topLeft.y;
    yBirdLoc = yBirdLoc + 1;
    return calculateBirdCoords(yBirdLoc);
}



