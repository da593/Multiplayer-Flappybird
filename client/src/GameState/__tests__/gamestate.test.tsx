import React  from 'react';
import { GameState } from 'GameState';
import renderer from 'react-test-renderer';


test("test game state for 1-player", () => {
    const component = renderer.create(<GameState players={1}/>)
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot()


})

