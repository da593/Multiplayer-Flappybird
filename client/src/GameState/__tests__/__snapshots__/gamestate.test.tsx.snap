// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test game state for 1-player 1`] = `
<div
  className="gameboard"
>
  <p
    className="score"
  >
    0
  </p>
  <canvas
    className="canvas-item"
    height={640}
    width={540}
  />
  <canvas
    className="canvas-item"
    height={640}
    width={540}
  />
  <canvas
    height={640}
    id="background-layer"
    width={540}
  />
</div>
`;
